File: cherokee/list.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*- */

/* Cherokee
 *
 * Authors:
 *      Alvaro Lopez Ortega <alvaro@alobbs.com>
 *
 * Copyright (C) 2001-2008 Alvaro Lopez Ortega
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of version 2 of the GNU General Public
 * License as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 * USA
 */

#include "common-internal.h"
#include "list.h"

#include <stdlib.h>

ret_t
cherokee_list_get_len (cherokee_list_t *head, size_t *len)
{
	cherokee_list_t *i;		
	cuint_t          n = 0;
	
	list_for_each (i, head) 
		n++;

	*len = n;
	return ret_ok;
}


void 
cherokee_list_sort (cherokee_list_t *head, int (*cmp)(cherokee_list_t *a, cherokee_list_t *b))
{
	cherokee_list_t *p, *q, *e, *list, *tail, *oldhead;
	int insize, nmerges, psize, qsize, i;
	
	list = head->next;
	cherokee_list_del(head);
	insize = 1;
	for (;;) {
		p = oldhead = list;
		list = tail = NULL;
		nmerges = 0;

		while (p) {
			nmerges++;
			q = p;
			psize = 0;
			for (i = 0; i < insize; i++) {
				psize++;
				q = q->next == oldhead ? NULL : q->next;
				if (!q)
					break;
			}

			qsize = insize;
			while (psize > 0 || (qsize > 0 && q)) {
				if (!psize) {
					e = q;
					q = q->next;
					qsize--;
					if (q == oldhead)
						q = NULL;
				} else if (!qsize || !q) {
					e = p;
					p = p->next;
					psize--;
					if (p == oldhead)
						p = NULL;
				} else if (cmp(p, q) <= 0) {
					e = p;
					p = p->next;
					psize--;
					if (p == oldhead)
						p = NULL;
				} else {
					e = q;
					q = q->next;
					qsize--;
					if (q == oldhead)
						q = NULL;
				}
				if (tail)
					tail->next = e;
				else
					list = e;
				e->prev = tail;
				tail = e;
			}
			p = q;
		}

		tail->next = list;
		list->prev = tail;

		if (nmerges <= 1)
			break;

		insize *= 2;
	}

	head->next = list;
	head->prev = list->prev;
	list->prev->next = head;
	list->prev = head;
}



ret_t 
cherokee_list_add_content (cherokee_list_t *head, void *item)
{
	CHEROKEE_NEW_STRUCT(n,list_item);
	
	/* Init
	 */
	INIT_LIST_HEAD (LIST(n));
	n->info = item;
	
	/* Add to list
	 */
	cherokee_list_add (LIST(n), head);
	
	return ret_ok;
}


ret_t 
cherokee_list_add_tail_content (cherokee_list_t *head, void *item)
{
	CHEROKEE_NEW_STRUCT(n,list_item);
	
	/* Init
	 */
	INIT_LIST_HEAD (LIST(n));
	n->info = item;
	
	/* Add to list
	 */
	cherokee_list_add_tail (LIST(n), head);
	
	return ret_ok;
}


ret_t 
cherokee_list_content_free (cherokee_list_t *head, cherokee_list_free_func free_func)
{
	cherokee_list_t *i, *tmp;

	list_for_each_safe (i, tmp, head) {
		cherokee_list_content_free_item (i, free_func);
	}
	
	INIT_LIST_HEAD(head);
	
	return ret_ok;
}


ret_t 
cherokee_list_content_free_item (cherokee_list_t *head, cherokee_list_free_func free_func)
{
	cherokee_list_del (head);
	
	if ((free_func != NULL) && (LIST_ITEM(head)->info)) {
		free_func (LIST_ITEM(head)->info);
	}
	
	free (head);
	return ret_ok;
}


ret_t 
cherokee_list_content_free_item_simple (cherokee_list_t *head)
{
	cherokee_list_del (head);
	
	if (LIST_ITEM(head)->info) {
		free (LIST_ITEM(head)->info);
	}
	
	free (head);
	return ret_ok;	
}

File: admin/wizards/List.py
# -*- coding: utf-8 -*-
#
# Cherokee-admin
#
# Authors:
#      Alvaro Lopez Ortega <alvaro@alobbs.com>
#
# Copyright (C) 2001-2014 Alvaro Lopez Ortega
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 2 of the GNU General Public
# License as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301, USA.
#

TYPE_VSERVER = 1
TYPE_RULE    = 1 << 2

LIST = [
    {'name':  'CMS',
     'title': N_('CMS'),
     'descr': N_('Content Management Systems'),
     'list':  [
            {'name':  'concrete5',
             'title': N_('Concrete 5'),
             'descr': N_('The CMS made for Marketing, but built for Geeks.'),
             'type':  TYPE_VSERVER | TYPE_RULE},
            ]
     },

    {'name':  'appservers',
     'title': N_('Application Servers'),
     'descr': N_('General Purpose Application Servers'),
     'list':  [
            {'name':  'alfresco',
             'title': N_('Alfresco'),
             'descr': N_('Enterprise web platform for building business solutions.'),
             'type':  TYPE_VSERVER | TYPE_RULE},
            {'name':  'glassfish',
             'title': N_('Glassfish'),
             'descr': N_('The first compatible implementation of the Java EE 6 platform specification.'),
             'type':  TYPE_VSERVER | TYPE_RULE},
            {'name':  'coldfusion',
             'title': N_('Adobe ColdFusion'),
             'descr': N_('Rapidly build Internet applications for the enterprise.'),
             'type':  TYPE_VSERVER | TYPE_RULE}
             ]
    },

    {'name':  'platforms',
     'title': N_('Platforms'),
     'descr': N_('Web Development Platforms'),
     'list':  [
            {'name':  'django',
             'title': N_('Django'),
             'descr': N_('The Web framework for perfectionists (with deadlines).'),
             'type':  TYPE_VSERVER | TYPE_RULE},
            {'name':  'rails',
             'title': N_('Ruby on Rails'),
             'descr': N_('Open-source web framework optimized for sustainable productivity.'),
             'type':  TYPE_VSERVER | TYPE_RULE},
            {'name':  'zend',
             'title': N_('Zend'),
             'descr': N_('Simple, straightforward, open-source software framework for PHP 5.'),
             'type':  TYPE_VSERVER | TYPE_RULE},
            {'name':  'symfony',
             'title': N_('Symfony'),
             'descr': N_('Full-stack framework, a library of cohesive classes written in PHP.'),
             'type':  TYPE_VSERVER | TYPE_RULE},
            {'name':  'uwsgi',
             'title': N_('uWSGI'),
             'descr': N_('Fast, self-healing, developer-friendly, language-agnostic, application container.'),
             'type':  TYPE_VSERVER | TYPE_RULE}
             ]
    },

    {'name':  'langs',
     'title': N_('Languages'),
     'descr': N_('Development Languages and Platforms'),
     'list':  [
            {'name':  'php',
             'title': N_('PHP'),
             'descr': N_('Widely-used general-purpose scripting language.'),
             'type':  TYPE_VSERVER | TYPE_RULE},
            {'name':  'mono',
             'title': N_('.NET with Mono'),
             'descr': N_('Cross-platform, open-source .NET development framework.'),
             'type':  TYPE_VSERVER | TYPE_RULE},
             ]
    },

    {'name':  'webapps',
     'title': N_('Web Applications'),
     'descr': N_('General Purpose Applications'),
     'list':  [
            {'name':  'mailman',
             'title': N_('GNU MailMan'),
             'descr': N_('Free software for managing electronic mail discussion and e-newsletter lists.'),
             'type':  TYPE_VSERVER},
            {'name':  'rtorrent',
             'title': N_('rTorrent'),
             'descr': N_('BitTorrent client written in C++, based on the libTorrent libraries for Unix.'),
             'type':  TYPE_RULE},
             ]
    },

    {'name':  'tasks',
     'title': N_('Tasks'),
     'descr': N_('Common Maintenance Tasks'),
     'list':  [
            {'name':  'ssl_test',
             'title': N_('SSL/TLS Testing'),
             'descr': N_('Auto-configure SSL/TLS support with a self-signed certificate.'),
             'type':  TYPE_RULE},
            {'name':  'flcache',
             'title': N_('Content Caching'),
             'descr': N_('Boost your server performance by enabling content caching.'),
             'type':  TYPE_RULE},
            {'name':  'redirect',
             'title': N_('Virtual Server Redirection'),
             'descr': N_('Create a new virtual server redirecting every request to another domain host.'),
             'type':  TYPE_VSERVER},
            {'name':  'hotlinking',
             'title': N_('Anti Hot-Linking'),
             'descr': N_('Stop other domains from hot-linking your media files.'),
             'type':  TYPE_RULE},
            {'name':  'icons',
             'title': N_('Directory-listing Icons'),
             'descr': N_('Add the /cherokee_icons and /cherokee_themes directories used for directory listing.'),
             'type':  TYPE_RULE},
            {'name':  'static',
             'title': N_('Static-file support'),
             'descr': N_('Add a rule to optimally serve the most common static files.'),
             'type':  TYPE_RULE},
            {'name':  'streaming',
             'title': N_('Media file streaming'),
             'descr': N_('Add a rule to stream media files through an intelligent streaming handler.'),
             'type':  TYPE_RULE},
             ]
    },
]
